File: sdks/node/test/client.test.js
import assert from "node:assert/strict";
import test from "node:test";

import { Notilify, NotilifyError } from "../src/index.js";

test("sends the canonical message request", async () => {
  let captured;
  const client = new Notilify("key_test", {
    fetch: async (url, options) => {
      captured = { url, options };
      return new Response(JSON.stringify({ status: true, message: "Message Accepted", data: { id: "msg_1" } }), { status: 202 });
    }
  });

  const result = await client.sendMessage({
    from: "NOTILIFY",
    to: "+14155552671",
    message: "Your code is 482913",
    idempotencyKey: "send-1"
  });

  assert.equal(captured.url, "https://api.notilify.com/v1/message");
  assert.equal(captured.options.headers.Authorization, "Bearer key_test");
  assert.equal(captured.options.headers["Idempotency-Key"], "send-1");
  assert.deepEqual(JSON.parse(captured.options.body), {
    from: "NOTILIFY",
    to: "+14155552671",
    message: "Your code is 482913"
  });
  assert.equal(result.data.id, "msg_1");
});

test("throws a structured API error", async () => {
  const client = new Notilify("key_test", {
    fetch: async () => new Response(JSON.stringify({ status: false, message: "Invalid sender", data: { field: "from" } }), { status: 400 })
  });

  await assert.rejects(
    client.sendMessage({ from: "BAD", to: "+14155552671", message: "Hello" }),
    (error) => error instanceof NotilifyError && error.statusCode === 400 && error.data.field === "from"
  );
});

File: sdks/ruby/test/client_test.rb
require "minitest/autorun"
require_relative "../lib/notilify"

Response = Struct.new(:code, :body)

class ClientTest < Minitest::Test
  def test_sends_canonical_message
    captured = nil
    transport = lambda do |uri, request|
      captured = [uri, request]
      Response.new("202", '{"status":true,"message":"Message Accepted","data":{"id":"msg_1"}}')
    end
    client = Notilify::Client.new("key_test", transport: transport)

    result = client.send_message(from: "NOTILIFY", to: "+14155552671", message: "Your code is 482913", idempotency_key: "send-1")

    assert_equal "/v1/message", captured[0].path
    assert_equal "Bearer key_test", captured[1]["Authorization"]
    assert_equal "send-1", captured[1]["Idempotency-Key"]
    assert_equal "msg_1", result.dig("data", "id")
  end

  def test_raises_structured_error
    transport = ->(*) { Response.new("400", '{"status":false,"message":"Invalid sender","data":{"field":"from"}}') }
    error = assert_raises(Notilify::Error) do
      Notilify::Client.new("key_test", transport: transport).send_message(from: "BAD", to: "+14155552671", message: "Hello")
    end
    assert_equal 400, error.status_code
    assert_equal "from", error.data["field"]
  end
end
